wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```
@@ -4977,21 +5063,72 @@ private function cast_value_in_non_strict_mode(
 				} elseif ( 'datetime' === $mysql_data_type || 'timestamp' === $mysql_data_type ) {
 					$function_call = sprintf( 'DATETIME(%s)', $translated_value );
 				} elseif ( 'year' === $mysql_data_type ) {
-					$function_call = sprintf( "STRFTIME('%%Y', %s)", $translated_value );
+					/*
+					 * The YEAR type in MySQL only uses 1 byte and therefore
+					 * covers only 256 values from 1901 to 2155 included.
+					 * Additionally:
+					 *   - Numbers from 0 to 69 correspond to years 2000 to 2069.
+					 *   - Numbers from 70 to 99 correspond to years 1970 to 1999.
+					 */
+					return sprintf(
+						"(
+							SELECT CASE
+								WHEN value IS NULL THEN NULL
+								WHEN value = 0 THEN '0000'
+								WHEN value BETWEEN 1901 AND 2155 THEN value
+								WHEN value BETWEEN 1 AND 69 THEN 2000 + value
+								WHEN value BETWEEN 70 AND 99 THEN 1900 + value
+								ELSE %s
+							END
+							FROM (SELECT CAST(%s AS INTEGER) AS value)
+						)",
+						$is_strict_mode
+							? sprintf( "THROW('Out of range value: ''' || %s || '''')", $translated_value )
+							: "'0000'",
+						$translated_value
+					);
 				}
 
-				// When the function call evaluates to NULL (invalid date/time),
-				// we need to fallback to the IMPLICIT DEFAULT value.
+				// In strict mode, invalid date/time values are rejected.
+				// In non-strict mode, they get an IMPLICIT DEFAULT value.
+				if ( $is_strict_mode ) {
+					$fallback = sprintf(
+						"THROW('Incorrect %s value: ''' || %s || '''')",
```

Such a nice use of callbacks 🤯 ! At the same time – could we just throw in place? Do we need to delay it? What's the benefit of doing this vs just throwing a new exception right here?

---

@adamziel We need to throw the error from SQLite, because in PHP we don't know the value. You can insert known values (`INSERT INTO ... VALUES (1, 2, 3)`), as well as unknown ones `INSERT INTO ... SELECT ...`. For that reason, we need SQLite to evaluate this for every inserted/updated row, as we can't do that in PHP. In PHP the "value" can be a column name, an expression, etc.